### PR TITLE
Feature/dirty flag btc 583

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/blockchain/block_index.hpp [hunk 1]
```diff
@@ -41,6 +41,8 @@ enum BlockStatus : uint8_t {
       BLOCK_FAILED_CHILD | BLOCK_FAILED_POP | BLOCK_FAILED_BLOCK,
   //! the block has been applied via PopStateMachine
   BLOCK_APPLIED = 1 << 5,
+  //! the block has been modified, added or removed
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:44]:
```suggestion
  //! the block has been modified, and should be written on disk
```

File: include/veriblock/blockchain/alt_block_addon.hpp [hunk 2]
```diff
@@ -25,94 +25,110 @@ struct AltBlockAddon : public PopState<AltEndorsement> {
...
+    return _atvids == o._atvids && _vtbids == o._vtbids &&
+           _vbkblockids == o._vbkblockids &&
            PopState<AltEndorsement>::operator==(o);
   }
 
   template <typename pop_t, typename pop_id_t>
-  std::vector<pop_id_t>& getPayloadIds();
+  const std::vector<pop_id_t>& getPayloadIds() const;
 
   template <typename pop_t, typename pop_id_t>
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:41]:
```suggestion
  template <typename pop_t>
```

File: include/veriblock/blockchain/alt_block_addon.hpp [hunk 3]
```diff
@@ -25,94 +25,110 @@ struct AltBlockAddon : public PopState<AltEndorsement> {
...
 
   bool operator==(const AltBlockAddon& o) const {
-    return atvids == o.atvids && vtbids == o.vtbids &&
-           vbkblockids == o.vbkblockids &&
+    return _atvids == o._atvids && _vtbids == o._vtbids &&
+           _vbkblockids == o._vbkblockids &&
            PopState<AltEndorsement>::operator==(o);
   }
 
   template <typename pop_t, typename pop_id_t>
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:38]:
```suggestion
  template <typename pop_t>
```

File: include/veriblock/blockchain/alt_block_addon.hpp [hunk 4]
```diff
@@ -25,94 +25,110 @@ struct AltBlockAddon : public PopState<AltEndorsement> {
...
-  std::vector<pop_id_t>& getPayloadIds();
+  const std::vector<pop_id_t>& getPayloadIds() const;
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:39]:
```suggestion
  const std::vector<typename pop_t::id_t>& getPayloadIds() const;
```

File: include/veriblock/blockchain/alt_block_addon.hpp [hunk 5]
```diff
@@ -25,94 +25,110 @@ struct AltBlockAddon : public PopState<AltEndorsement> {
...
 
   template <typename pop_t, typename pop_id_t>
-  const std::vector<pop_id_t>& getPayloadIds() const;
+  void removePayloadIds(const pop_id_t& pid) {
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:42]:
```suggestion
  void removePayloadIds(const typename pop_t::id_t& pid) {
```

File: include/veriblock/blockchain/alt_block_addon.hpp [hunk 6]
```diff
@@ -25,94 +25,110 @@ struct AltBlockAddon : public PopState<AltEndorsement> {
...
+  }
 
-  void setNull() {
-    PopState<AltEndorsement>::setNull();
-    chainWork = 0;
-    atvids.clear();
-    vtbids.clear();
-    vbkblockids.clear();
+  template <typename pop_t, typename pop_id_t>
+  void insertPayloadIds(const pop_id_t& pid) {
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:50]: same

Comment (Warchant) [inline include/veriblock/blockchain/alt_block_addon.hpp:50]: Either make this method singular, e.g. `insertPayloadId` or accept vector of ids (or both, as 2 separate methods)

File: include/veriblock/blockchain/block_index.hpp [hunk 7]
```diff
@@ -91,19 +86,44 @@ struct BlockIndex : public Block::addon_t {
...
 
+  height_t getHeight() const { return height; }
+  void setHeight(const height_t newHeight) {
+    height = newHeight;
+    setDirty();
+  }
+
+  const block_t& getHeader() const { return *header; }
+  void setHeader(const block_t& newHeader) {
+    header = std::make_shared<block_t>(newHeader);
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:121]: add another overload, which accepts shared_ptr and moves it to member field

File: src/blockchain/alt_block_tree.cpp [hunk 8]
```diff
@@ -76,7 +76,7 @@ bool payloadsCheckDuplicates(Index& index,
 template <typename Index, typename Pop>
 void commitPayloadsIds(Index& index, const std::vector<Pop>& pop) {
   for (const auto& p : pop) {
-    index.template insertPayloadIds<Pop, typename Pop::id_t>(p.getId());
+    index.template insertPayloadId<Pop>(p.getId());
   }
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:80]: maybe add method `insertPayloadsIds<Pop>(pop)`? inserting vector into another is more effective than just inserting same vector 1by1

File: include/veriblock/blockchain/base_block_tree.hpp [hunk 9]
```diff
@@ -355,20 +355,20 @@ struct BaseBlockTree {
     VBK_ASSERT(header != nullptr);
 
     index_t* current = touchBlockIndex(header->getHash());
-    current->header = header;
-    current->pprev = getBlockIndex(header->previousBlock);
+    current->setHeader(*header);
```
Comment (Warchant) [inline include/veriblock/blockchain/base_block_tree.hpp:358]: here, use `setHeader` which accepts pointer, e.g.
```
current->setHeader(std::move(header));
```

File: include/veriblock/blockchain/block_index.hpp [hunk 10]
```diff
@@ -197,6 +224,16 @@ struct BlockIndex : public Block::addon_t {
...
   bool operator!=(const BlockIndex& b) const { return !operator==(b); }
+
+ protected:
+  //! height of the entry in the chain
+  height_t height = 0;
+
+  //! block header
+  std::shared_ptr<block_t> header{};
+
+  //! contains status flags
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:235]:
```suggestion
  //! (memory-only) contains status flags
```

File: include/veriblock/blockchain/pop/pop_state.hpp [hunk 11]
```diff
@@ -19,43 +19,58 @@ struct PopState {
...
   //! (memory-only) list of endorsements pointing to this block
   std::vector<endorsement_t*> endorsedBy;
 
+  const std::unordered_map<eid_t, std::shared_ptr<endorsement_t>>&
+    getContainingEndorsements() const {
+    return _containingEndorsements;
+  }
+
+  void insertContainingEndorsement(std::shared_ptr<endorsement_t> e) {
+    _containingEndorsements.insert(std::make_pair(e->id, e));
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_state.hpp:31]:
```suggestion
    _containingEndorsements.insert(std::make_pair(e->id, std::move(e)));
```

File: pypopminer/src/entities.cpp [hunk 12]
```diff
@@ -119,27 +119,28 @@ void init_entities() {
...
+      .def("getHeight", &VbkBlock::getHeight)
+      .def("getVersion", &VbkBlock::getVersion)
+      .def("getPreviousBlock", &VbkBlock::getPreviousBlock)
+      .def("getPreviousKeystone", &VbkBlock::getPreviousKeystone)
+      .def("getSecondPreviousKeystone", &VbkBlock::getSecondPreviousKeystone)
+      .def("getMerkleRoot", &VbkBlock::getMerkleRoot)
+      .def("setMerkleRoot", &VbkBlock::setMerkleRoot)
+      .def("getBlockTime", &VbkBlock::getBlockTime)
+      .def("getDifficulty", &VbkBlock::getDifficulty)
+      .def("getNonce", &VbkBlock::getNonce);
```
Comment (Warchant) [inline pypopminer/src/entities.cpp:145]: please do not change anything in mock miner

File: include/veriblock/blockchain/block_index.hpp (hunk 10 above)
Comment (blackyblack) [reply include/veriblock/blockchain/block_index.hpp:235]: It is not true yet.

**Pull request merged**